File: src/main/java/cn/xihan/lib/easyhttp/callback/BaseCallback.java
package cn.xihan.lib.easyhttp.callback;


import cn.xihan.lib.easyhttp.EasyConfig;
import cn.xihan.lib.easyhttp.EasyLog;
import cn.xihan.lib.easyhttp.EasyUtils;
import cn.xihan.lib.easyhttp.model.CallProxy;
import cn.xihan.lib.easyhttp.request.HttpRequest;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.SocketTimeoutException;

/**
 *    author : Android 轮子哥
 *    github : <a href="https://github.com/getActivity/EasyHttp">...</a>
 *    time   : 2019/11/25
 *    desc   : 接口回调基类
 */
public abstract class BaseCallback implements Callback {

    /** 请求配置 */
    private final HttpRequest<?> mHttpRequest;

    /** 请求任务对象 */
    private CallProxy mCall;

    /** 当前重试次数 */
    private int mRetryCount;

    public BaseCallback(HttpRequest<?> request) {
        mHttpRequest = request;
    }

    public BaseCallback setCall(CallProxy call) {
        mCall = call;
        return this;
    }

    public void start() {
        mCall.enqueue(this);
        onStart(mCall);
    }

    protected CallProxy getCall() {
        return mCall;
    }

    @Override
    public void onResponse(@NotNull Call call, @NotNull Response response) {
        try {
            // 收到响应
            onResponse(response);
        } catch (Exception e) {
            // 回调失败
            onFailure(e);
        } finally {
            // 关闭响应
            closeResponse(response);
        }
    }

    @Override
    public void onFailure(@NotNull Call call, @NotNull IOException e) {
        // 服务器请求超时重试
        if (e instanceof SocketTimeoutException && mRetryCount < EasyConfig.getInstance().getRetryCount()) {
            // 设置延迟 N 秒后重试该请求
            EasyUtils.postDelayed(() -> {

                mRetryCount++;
                Call newCall = call.clone();
                mCall.setCall(newCall);
                newCall.enqueue(BaseCallback.this);
                // 请求超时，正在执行延迟重试
                EasyLog.printLog(mHttpRequest, "The request timed out, a delayed retry is being performed, the number of retries: " +
                        mRetryCount + " / " + EasyConfig.getInstance().getRetryCount());

            }, EasyConfig.getInstance().getRetryTime());

            return;
        }
        onFailure(e);
    }

    /**
     * 请求开始
     */
    protected abstract void onStart(Call call);

    /**
     * 请求成功
     */
    protected abstract void onResponse(Response response) throws Exception;

    /**
     * 请求失败
     */
    protected abstract void onFailure(Exception e);

    /**
     * 关闭响应
     */
    protected void closeResponse(Response response) {
        EasyUtils.closeStream(response);
    }
}

File: src/main/java/cn/xihan/lib/easyhttp/callback/BaseCallback.kt
package cn.xihan.lib.easyhttp.callback

import cn.xihan.lib.easyhttp.EasyConfig
import cn.xihan.lib.easyhttp.EasyLog.printLog
import cn.xihan.lib.easyhttp.EasyUtils.closeStream
import cn.xihan.lib.easyhttp.EasyUtils.postDelayed
import cn.xihan.lib.easyhttp.model.CallProxy
import cn.xihan.lib.easyhttp.request.HttpRequest
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import java.io.IOException
import java.net.SocketTimeoutException

/**
 * author : Android 轮子哥
 * github : [https://github.com/getActivity/EasyHttp](https://github.com/getActivity/EasyHttp)
 * time   : 2019/11/25
 * desc   : 接口回调基类
 */
abstract class BaseCallback(
    /** 请求配置  */
    private val mHttpRequest: HttpRequest<*>
) : Callback {
    /** 请求任务对象  */
    protected var call: CallProxy? = null
        private set

    /** 当前重试次数  */
    private var mRetryCount = 0
    fun setCall(call: CallProxy?): BaseCallback {
        this.call = call
        return this
    }

    open fun start() {
        call!!.enqueue(this)
        onStart(call)
    }

    override fun onResponse(call: Call, response: Response) {
        try {
            // 收到响应
            onResponse(response)
        } catch (e: Exception) {
            // 回调失败
            onFailure(e)
        } finally {
            // 关闭响应
            closeStream(response)
        }
    }

    override fun onFailure(call: Call, e: IOException) {
        var call = call
        // 服务器请求超时重试
        if (e is SocketTimeoutException && mRetryCount < EasyConfig.getInstance().getRetryCount()) {
            // 设置延迟 N 秒后重试该请求
            postDelayed({
                mRetryCount++
                val newCall = call.clone()
                call = newCall
                newCall.enqueue(this@BaseCallback)
                // 请求超时，正在执行延迟重试
                printLog(
                    mHttpRequest, "The request timed out, a delayed retry is being performed, the number of retries: " +
                            mRetryCount + " / " + EasyConfig.getInstance().getRetryCount()
                )
            }, EasyConfig.getInstance().getRetryTime())
            return
        }
        onFailure(e)
    }

    /**
     * 请求开始
     */
    protected abstract fun onStart(call: Call?)

    /**
     * 请求成功
     */
    @Throws(Exception::class)
    protected abstract fun onResponse(response: Response?)

    /**
     * 请求失败
     */
    protected abstract fun onFailure(e: Exception)
}